src/fw-logs/fw-logs-parser.cpp

```diff
@@ -1,90 +1,315 @@
+            structured_data.file_id = static_cast< uint32_t >( log_binary->file_id );
+            structured_data.module_id = static_cast< uint32_t >( log_binary->module_id );
+            structured_data.event_id = static_cast< uint32_t >( log_binary->event_id );
+            structured_data.line = static_cast< uint32_t >( log_binary->line_id );
+            structured_data.sequence = static_cast< uint32_t >( log_binary->seq_id );
 
-            //parse second DWORD
-            log_data._event_id = static_cast<uint32_t>(log_binary->dword2.bits.event_id);
-            log_data._line = static_cast<uint32_t>(log_binary->dword2.bits.line_id);
-            log_data._sequence = static_cast<uint32_t>(log_binary->dword2.bits.seq_id);
+            return structured_data;
+        }
 
-            //parse third DWORD
-            log_data._p1 = static_cast<uint32_t>(log_binary->dword3.p1);
-            log_data._p2 = static_cast<uint32_t>(log_binary->dword3.p2);
+        void fw_logs_parser::structure_timestamp( const fw_logs_binary_data * raw,
+                                                  structured_binary_data * structured ) const
+        {
+            const auto actual_struct = reinterpret_cast< const extended_fw_log_binary * >( raw->logs_buffer.data() );
+            structured->timestamp = actual_struct->soc_timestamp;
+        }
 
-            //parse forth DWORD
-            log_data._p3 = static_cast<uint32_t>(log_binary->dword4.p3);
+        void fw_logs_parser::structure_params( const fw_logs_binary_data * raw,
+                                               size_t num_of_params,
+                                               structured_binary_data * structured ) const
+        {
+            const auto actual_struct = reinterpret_cast< const extended_fw_log_binary * >( raw->logs_buffer.data() );
+            if( num_of_params != actual_struct->number_of_params )
+                throw librealsense::invalid_value_exception( rsutils::string::from() << "Expecting " << num_of_params
+                                                             << " parameters, received " << actual_struct->number_of_params );
 
-            //parse fifth DWORD
-            log_data._timestamp = log_binary->dword5.timestamp;
+            if( num_of_params > 0 )
+            {
+                struct extended_fw_log_params_binary : public extended_fw_log_binary
+                {
+                    // param_info array size namber_of_params, 1 is just a placeholder. Need to use an array not a
+                    // pointer because pointer can be 8 bytes of size and than we won't parse the data correctly
+                    param_info info[1];
+                };
 
-            log_data._delta = (_last_timestamp == 0) ? 
-                0 :(log_data._timestamp - _last_timestamp) * _timestamp_factor;
+                const auto with_params = reinterpret_cast< const extended_fw_log_params_binary * >( raw->logs_buffer.data() );
```

Comment:
why the line:
reinterpret_cast< const extended_fw_log_params_binary * >( raw->logs_buffer.data() )
cannot be replaced by:
reinterpret_cast< const param_info * >( raw->logs_buffer.data() )

Reply:
Because param_info has offset relative to start of buffer. Other fields are inherited from `extended_fw_log_binary` and `with_params->info` points to the correct byte.

Initially I had this field in the `extended_fw_log_binary` definition. However it is an optional field and only used here, when parsing params, so I split it to another struct. This simplified my code in other places.